proposals/safe-fixed-size-buffers.md

```diff
@@ -0,0 +1,558 @@
+Safe Fixed Size Buffers
+=====
+
+## Summary
+
+Provide a general-purpose and safe mechanism for declaring fixed sized buffers within C# classes, structs, and interfaces.
+
+## Motivation
+
+This proposal plans to address the many limitations of https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/unsafe-code.md#228-fixed-size-buffers.
+Specifically it aims to allow the declaration of safe `fixed` buffers for managed and unmanaged types in a `struct`, `class`, or `interface`, and provide language safety verification for them.
+
+Note, that for the purpose of this proposal a term "fixed-size buffer" refers to a "safe fixed-size buffer" rather than to a buffer described at https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/unsafe-code.md#228-fixed-size-buffers.
+
+## Detailed Design (Option 1)
+
+The grammar for *variable_declarator* in https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/classes.md#145-fields
+will be extended to allow specifying the size of the buffer:
+
+``` diff antlr
+field_declaration
+    : attributes? field_modifier* type variable_declarators ';'
+    ;
+
+field_modifier
+    : 'new'
+    | 'public'
+    | 'protected'
+    | 'internal'
+    | 'private'
+    | 'static'
+    | 'readonly'
+    | 'volatile'
+    | unsafe_modifier   // unsafe code support
+    ;
+
+variable_declarators
+    : variable_declarator (',' variable_declarator)*
+    ;
+    
+variable_declarator
+    : identifier ('=' variable_initializer)?
++   | fixed_size_buffer_declarator
+    ;
+    
+fixed_size_buffer_declarator
+    : identifier '[' constant_expression ']'
+    ;    
+```
+
+A *fixed_size_buffer_declarator* introduces a new field and consists of an identifier that names it,
+followed by a constant expression enclosed in `[` and `]` tokens. The type of the constant expression must be
+implicitly convertible to type `int`, and the value must be a non-zero positive integer.
+
+A *field_declaration* with a *fixed_size_buffer_declarator* in an interface must have `static` modifier.
+
+The field has a special struct type which represents a fixed-size buffer of a given element type.
+The element type is the *type* specified in *field_declaration*.  The constant expression
+denotes the number of elements in the buffer. The elements of a fixed-size buffer are laid out sequentially
+in memory as though they are elements of an array.
+
+The struct type will be defined using [```System.Runtime.CompilerServices.InlineArrayAttribute```](https://github.com/dotnet/runtime/issues/61135) feature.
+The framework will likely include a set of predefined "buffer" types that cover a limited set of buffer sizes. When a predefined type doesn't exist,
```

Comment:
What is the motivation for exposing predefined "buffer" types from the BCL? Is it just to save a bit of binary size or is it more than that?

Reply:
> What is the motivation for exposing predefined "buffer" types from the BCL? Is it just to save a bit of binary size or is it more than that?

The primary motivation is to avoid synthesizing types for common sizes in many user assemblies. There is also a benefit of interop between different user assemblies. If, for example, a buffer of size 10 type is in BCL, all user assemblies can easily pass instance of the same type around without a need to copy data between different versions of buffers of size 10.